var/spack/repos/builtin/packages/py-pyarrow/package.py

```diff
@@ -36,39 +37,54 @@ class PyPyarrow(PythonPackage, CudaPackage):
     version("0.11.0", sha256="07a6fd71c5d7440f2c42383dd2c5daa12d7f0a012f1e88288ed08a247032aead")
     version("0.9.0", sha256="7db8ce2f0eff5a00d6da918ce9f9cfec265e13f8a119b4adb1595e5b19fd6242")
 
-    depends_on("cxx", type="build")  # generated
-
-    variant("parquet", default=False, description="Build with Parquet support")
-    variant("orc", default=False, description="Build with orc support")
-    variant("dataset", default=False, description="Build with Dataset support")
-
-    conflicts("~parquet", when="+dataset")
-
-    depends_on("cmake@3.0.0:", type="build")
-    depends_on("pkgconfig", type="build")
-    depends_on("python@3.8:", type=("build", "run"), when="@13:")
-    depends_on("python@3.7:", type=("build", "run"), when="@7:")
-    depends_on("python@3.6:", type=("build", "run"), when="@3:")
-    depends_on("python@3.5:", type=("build", "run"), when="@0.17:")
-    depends_on("py-setuptools", type="build")
-    depends_on("py-setuptools@40.1.0:", type="build", when="@10.0.1:")
-    depends_on("py-setuptools@38.6.0:", type="build", when="@7:")
-    depends_on("py-setuptools-scm@:7", type="build", when="@0.15:")
-    depends_on("py-cython", type="build")
-    depends_on("py-cython@0.29.31:", type="build", when="@14:")
-    depends_on("py-cython@0.29.31:2", type="build", when="@12:13")
-    depends_on("py-cython@0.29.22:2", type="build", when="@8:11")
-    depends_on("py-cython@0.29:2", type="build", when="@0.15:7")
-    depends_on("py-cython@:2", type="build", when="@:0.14")
-    # in newer pip versions --install-option does not exist
-    depends_on("py-pip@:23.0", type="build")
-
-    depends_on("py-numpy@1.16.6:", type=("build", "run"), when="@3:")
-    # Prior to python 3.9 numpy must be >=0.14,<1.25
-    depends_on("py-numpy@0.14:1.24", when="^python@:3.8", type=("build", "run"))
-    depends_on("py-numpy@1.25:", when="^python@3.9:", type=("build", "run"))
-    # https://github.com/apache/arrow/issues/39532
-    depends_on("py-numpy@:1", when="@:15", type=("build", "run"))
+    # Starting with pyarrow 17+, backend support is built if arrow was built with it
+    with when("@:16"):
+        variant("parquet", default=False, description="Build with Parquet support")
+        variant("orc", default=False, description="Build with orc support")
+        variant("dataset", default=False, description="Build with Dataset support")
```

Comment:
This is the kind of tricky variant deprecation problem. Someone might have `py-pyarrow +parquet`, and this PR will now pin them to 16 (potentially even downgrade from 18).  Should we approach this as a variant that remains valid but with a conflict for the false option when a recent version and using arrow with the true option?
```suggestion
    variant("parquet", default=False, description="Build with Parquet support")
    variant("orc", default=False, description="Build with orc support")
    variant("dataset", default=False, description="Build with Dataset support")
    with when("@17:"):
        conflicts("~parquet ^arrow +parquet")
        conflicts("~orc ^arrow +orc")
```

Reply:
Personally I would be in favor of dropping the variant altogether. There has never been any extras registered in the library, so `pip install pyarrow[parquet]` has never been supported. If we want to maintain intuitive compatibility with pip, we should get rid of the variant wherever it is referenced. We can get rid of it for older versions to if we condition it on `arrow` instead of `py-pyarrow`.

Reply:
Agreed, consistency with pip is best.

Reply:
Agreed, consistency with pip is best.